common/parallel/TreeCrawler.go

```diff
@@ -101,29 +154,117 @@ func (c *crawler) workerLoop(ctx context.Context, wg *sync.WaitGroup, workerInde
+		// and cause memory pressure on system.
+		for len(c.unstartedDirs) > maxQueueDirectories {
+			time.Sleep(10 * time.Millisecond)
+		}
+		c.cond.L.Lock()
+		if tDir != nil {
+			c.unstartedDirs = append(c.unstartedDirs, tDir)
+		}
+		c.cond.L.Unlock()
+		c.cond.Broadcast()
+	}
+
+	// There is no way to know if channel is closed or not without interacting with channel.
+	// So, set this flag for target traverser to know source traverser done.
+	c.cond.L.Lock()
+	c.mayHaveMoreDirs = false
+	c.cond.L.Unlock()
+	c.cond.Broadcast()
+}
+
+// autoPacerWait do the job of pacemaker between source and target traverser.
+func (c *crawler) autoPacerWait(ctx context.Context) {
+	const bytesInGB = 1024 * 1024 * 1024
+	//
+	// Consider 80% full as low-water-mark. Anything less than 80% full is fair game and we let the source traverser
+	// run full throttle. Once it exceeds low-water-mark we slowly start taking the foot off the pedal and every thread/goroutine
+	// needs to wait for a small time before proceeding. Once it touches/crosses high-water-mark no scanner thread/goroutine will
+	// be allowed to proceed. For reasonable sized directories this should ensure that we remain in MaxObjectIndexerSizeInBytes limit
+	// for most practical scenarios. For large/huge directories we might exceed MaxObjectIndexerSizeInBytes by the size of the largest directory.
+	//
+	MaxObjectIndexerSizeInBytes := int64(c.maxObjectIndexerSizeInGB * bytesInGB)
+	lowWaterMark := (MaxObjectIndexerSizeInBytes * 8) / 10
+
+	highWaterMark := MaxObjectIndexerSizeInBytes
+
+	mapSizeInBytes := c.getObjectIndexerMapSize()
+
+	// Nice sunny morning, press that pedal more.
+	if mapSizeInBytes < lowWaterMark {
+		return
+	}
+
+	//
+	// We are nearing danger zone, start slowing down.
+	//
+	// TODO: The sleep duration may be decided based on how fast the
+	//       objectIndexerMap is seen to be growing.
+	//
+	if mapSizeInBytes < highWaterMark && ctx.Err() == nil {
+		time.Sleep(1 * time.Second)
+		return
+	}
+
+	// In danger zone. Don't proceed any further without dropping the objectIndexerMap size below lowWaterMark.
+	for mapSizeInBytes > lowWaterMark && ctx.Err() == nil {
+		time.Sleep(1 * time.Second)
+		mapSizeInBytes = c.getObjectIndexerMapSize()
+	}
+
+	return
+}
+
+func (c *crawler) processOneDirectoryWithAutoPacer(ctx context.Context, workerIndex int) (bool, error) {
 	const maxQueueDirectories = 1000 * 1000
 	const maxQueueDirsForBreadthFirst = 100 * 1000 // figure is somewhat arbitrary.  Want it big, but not huge
 
 	var toExamine Directory
 	stop := false
 
+	// Before picking a new directory to enumerate, call autoPacerWait() to check if we need to
+	// slow down as the objectIndexer might be getting "too full". If Target Traverser is running
+	// slow causing objectIndexer map to get full, we induce wait to let Target Traverser catch up.
+	// autoPacerWait is applicable in case if it's sync and source traverser.
```

Comment:
// Only source traverser of the sync process should call auto-pacer to slow itself down.